core/nightshade/src/nightshade_task.rs

```diff
@@ -0,0 +1,412 @@
+
+        // Check we already have such payload and request it otherwise
+        if let Some(signed_block) = &self.authority_blocks[author] {
+            if signed_block.block.hash() == message.state.block_hash() {
+                self.nightshade_as_mut_ref().update_state(message.sender_id, message.state);
+            } else {
+                self.nightshade_as_mut_ref().set_adversary(author);
+            }
+        } else {
+            // TODO: This message is discarded if we haven't receive the payload yet. We can store it
+            // in a queue instead, and process it after we have the payload.
+
+            let gossip = Gossip::new(
+                self.owner_id,
+                author,
+                GossipBody::PayloadRequest(vec!(author)),
+                &self.owner_secret_key,
+            );
+            self.send_gossip(gossip);
+        }
+    }
+
+    fn process_gossip(&mut self, gossip: Gossip<P>) {
+        if !gossip.verify(&self.public_keys[gossip.sender_id]) {
+            return;
+        }
+
+        match gossip.body {
+            GossipBody::NightshadeStateUpdate(message) => self.process_message(*message),
+            GossipBody::PayloadRequest(authorities) => self.send_payloads(gossip.sender_id, authorities),
+            GossipBody::PayloadReply(payloads) => self.receive_payloads(gossip.sender_id, payloads),
+        }
+    }
+
+    fn send_payloads(&self, receiver_id: AuthorityId, authorities: Vec<AuthorityId>) {
+        let mut payloads = Vec::new();
+        for a in authorities {
+            if let Some(ref p) = self.authority_blocks[a] {
+                payloads.push(p.clone());
+            }
+        }
+        let gossip = Gossip::new(
+            self.owner_id,
+            receiver_id,
+            GossipBody::PayloadReply(payloads),
+            &self.owner_secret_key,
+        );
+        self.send_gossip(gossip);
+    }
+
+    fn receive_payloads(&mut self, sender_id: AuthorityId, payloads: Vec<SignedBlock<P>>) {
+        for signed_payload in payloads {
+            let authority_id = signed_payload.block.author();
+
+            // If the signed block is not properly signed by its author, we mark the sender as adversary.
+            if !signed_payload.verify(&self.public_keys[authority_id]) {
+                self.nightshade_as_mut_ref().set_adversary(sender_id);
+                continue;
+            }
+
+            if let Some(ref p) = self.authority_blocks[authority_id] {
+                if p.block.hash() != signed_payload.block.hash() {
+                    self.nightshade_as_mut_ref().set_adversary(authority_id);
+                    self.authority_blocks[authority_id] = None;
+                }
+            } else {
+                self.authority_blocks[authority_id] = Some(signed_payload);
+                self.missing_payloads -= 1;
+            }
+        }
+    }
+
+    /// Sends gossip to random authority peers.
```

Comment:
Looks like it sends to everyone and not just random sample.

Reply:
This comment is deprecated as you notice, I will remove it though it will be a good idea not to send each update to every participant and design a better underlying gossip mechanism regarding state updates. I'll put this in Nomicon too.